src/main/scala/org/scalacheck/Prop.scala

```diff
@@ -804,81 +802,81 @@ object Prop {
 
   }
 
-  /** Universal quantifier for an explicit generator. Shrinks failed arguments
-   *  with the default shrink function for the type */
+  /** Universal quantifier for an explicit generator. */
+  @deprecated("Use the variant without Shrink.  Here only for bincompat.", "1.15.0")
```

Comment:
How does one still use shrinking? In QuickCheck, the default for `forAll` is no shrinking, but there is still support for shrinking with `forAllShrink`?  From what I see so far, that isn't available here.  This reads like turning off shrinking entirely, when the proposal was changing the default.

Reply:
> How does one still use shrinking?

It's still there, but _only_ for the variants of the `forAll` function with the implicit `Arbitrary`. I'm proposing to disable it for the `forAll` functions that take `Gen`.

[def forAll[A1,P] (f: A1 => P)(implicit p: P => Prop, a1: Arbitrary[A1], s1: Shrink[A1]): Prop](https://github.com/rickynils/scalacheck/blob/86fdea6aaebbdf8163324a690f3abdd5fa92593d/src/main/scala/org/scalacheck/Prop.scala#L1012)

Reply:
>  In QuickCheck, the default for forAll is no shrinking

Do you mean ScalaCheck? QuickCheck has always only [shrunk with Arbitrary](https://github.com/rickynils/scalacheck/blob/136f7df0b5c3b184ec434dbf6ecc4c5977f91ced/src/main/scala/org/scalacheck/Prop.scala#L900) ([if implemented for a specific instance](http://hackage.haskell.org/package/QuickCheck-2.12.4/docs/src/Test.QuickCheck.Arbitrary.html#shrink)) and [_never_ with Gen](http://hackage.haskell.org/package/QuickCheck-2.12.4/docs/src/Test.QuickCheck.Property.html#forAll), which is what I'm proposing to do in this PR.

Reply:
Maybe I missed the nuance of the change then.

So you're proposing four cases:

1) `forAll` and explicit `Gen`s with shrinking disabled
3) `forAll` with implicit `Arbitrary`s with shrinking preserved
3) `forAllShrink` and explicit `Gen`s with shrinking preserved
4) `forAllNoShrink` with implicit `Arbitrary`s

And still supporting all the possible arities of each.

Right now, ScalaCheck does:

1) `forAll` and explicit `Gen`s with shrinking
3) `forAll` with implicit `Arbitrary`s with shrinking
3) `forAllNoShrink` with explicit `Gen`s with shrinking
4) `forAllNoShrink` with implicit `Arbitrary`s

Reply:
> So you're proposing four cases:

Yeah something like that. The `forAllShrink` already exists, but takes a manual shrink function. I'm not touching that. I'm only changing the behaviour of `forAll` with `Gen` which is where the problem is. Everything else stays the same.

I suspect what I've proposed at the moment might have to be tweaked based on @rickynils comment and my own counter-question below (which is totally fine).

Reply:
I sense this will be a problem.  There needs to be a `forAllShrink` with an implicit 
 `Shrink`.  I don't entirely grok the ScalaCheck internals, but I believe the existing `forAllShrink` is not of any help here.

Reply:
> I believe the existing forAllShrink is not of any help here.

As a convenient function for taking X number of `Gen`s with implicit shrinking you're quite right. The current function requires the manual shrinking and only a single Gen.

> There needs to be a forAllShrink with an implicit Shrink

On this I'm afraid I don't agree. I'm curious who is building/using manual `Gen` with a restricted subset of the domain (eg. a subset of `Int`, say between 100 and 200) and still want implicit shrinking that might completely go-outside of that domain (eg. shrink to 0)? How do you know what the global/implicit `Shrink` is going to do and whether it aligns with your custom `Gen` range?

I hope this doesn't sound too lazy, but I don't think I would want to add that missing set of functions unless someone _really_ demanded/wanted it. I would try to convince them first it's a bad/dangerous idea. They can still use `Arbitrary` for their own types with their own `Shrink` implementations. :)

Reply:
> I'm curious who is building/using manual `Gen` with a restricted subset of the domain and still want implicit shrinking?

It's a flexibility issue.  Users could be declaring a `Gen` to `forAllShrink` but still defining their own `Shrink` definition in implicit scope, [which is supported in ScalaCheck](https://github.com/rickynils/scalacheck/blob/f31c3ed/doc/UserGuide.md#test-case-minimisation).

Reply:
> Users could be declaring a Gen to forAllShrink but still defining their own Shrink definition in implicit scope,

```scala
implicit def ShrinkMyInt: Shrink[Int] = ... // Shrink towards 10 say

//  This will shrink the second generator incorrectly
forAll(Gen.choose(10, 20), Gen.choose(30, 40)) { ... }
```

In the above example you can only provide _one_ implicit for `Shrink[Int]` in scope, but you really need one for _each_ `Gen` otherwise you're just asking for the same kinds of bugs/errors that this PR is trying to fix. I would argue you really shouldn't be doing anything implicit with `Shrink` if you're providing explicit `Gen`. Now perhaps we might want _explicit_ Gen _and_ Shrink together, which is totally fine...

```scala
forAllShrink[A, B](g1: Gen[A], s1: Shrink[A], g2: Gen[B], s2: Shrink[B])(....)
```

At this point I might suggest making a case class with Gen + Shrink to make it easier to pair/keep them together though.

```scala
case class GenShrink[A](gen: Gen[A], shrink: Shrink[A]) 
```

Reply:
There are ways to use different scopes, but supporting generatorso with shrinking having is a usability issue.  Users need the flexibility to study their failing tests by inserting a more limited generator when they are working on an arbitrary property.

For instance, while writing an arbitrary property:

```scala
    forAllShrink { s: String =>
      ...
```

You may want to study the issue with a specific type of generator without giving up shrinking:

```scala
    forAllShrink(Gen.asciiStr) { s: String =>
      ...
```

Reply:
> There are ways to use different scopes

Short of passing in each shrinking manually I don't know how you solve the problem of having two generators of the same type (`Char` in this case):

```scala
forAllShrink(Gen.asciiStr, Gen.alphaNumChar) { (a, b) =>
```

There is no shrinker that will work (correctly) for both those cases, so you would need to pass them in manually:

```scala
forAllShrink
  (Gen.asciiStr, Gen.alphaNumChar)
  (Shrink.asciiStr, Shrink.alphaNumChar)
  { (a, b) =>
```

As soon as you pass in one shrink implicit you also have to pass in the extra `A => Pretty` implicit argument as well, which becomes extremely painful. As I said in my previous comment, if you want to keep supporting this you should pass them in explicitly (no implicitly), and I would be tempted to put them together.

```scala
forAllShrink(
    Gen.asciiStr, Shrink.asciiStr
  , Gen.alphaNumChar, Shrink.alphaNumChar
  ) { (a, b) =>
```

Reply:
> There are ways to use different scopes

Sorry we probably should have taken this to Gitter at some point. In summary, just to draw a line under this thread.

1. I don't agree that shrinking should _ever_ be "implicit" for Gen (only Arbitrary).
2. I'm more than happy to add whatever extra functionality that @rickynils would like to see, but given that I disagree with Shrinking being implicit for Gen, I'm not likely to add anything along those lines extra unless he asks.

Let me know if you want to continue this conversation on Gitter or some other medium. :)

Reply:
> I don't agree that shrinking should _ever_ be "implicit" for Gen (only Arbitrary).

This seems like a hard line to draw, but we can set it aside.  Instead, given that`forAll` with `Gen` in ScalaCheck presently supports shrinking, it seems it should still be available to users.
And it can be available without interfering with the `forAll` change proposed here, by having it be named `forAllShrink`.

Reply:
`forAllYoloShrink`? :trollface:

I have no standing, but I guess it would be fine (and maybe a good migration path) to have a forAll that takes an explicit Gen and an implicit Shrink.

Reply:
There are already is one.  We're talking about when arity is greater than one.

Reply:
I think we're on the same page; by "a forAll" I meant a variant set of forAll's.

Reply:
Ok, but I'm not sure how useful explicit shrinks would be.  I'd say implicits are a feature here.

Reply:
FWIW I've been working around this issue for years by using `implicit val noShrink: Shrink[MyType] = Shrink.shrinkAny` (couldn't use `forAllNoShrink` because of scalatest).

I don't think `Gen` with implicit `Shrink`  makes any sense. But I like the idea of combining them. Then the usability issue could be solved by adding a `withShrink(implicit shrink: Shrink[A])` method on `Gen`.

Reply:
> Then the usability issue could be solved by adding a withShrink(implicit shrink: Shrink[A]) method on Gen.

I'm curious. I hope you don't mind me asking. What `Gen`s are you constructing/using for your code base, and in particular what is the range of the values? And in what circumstances would you want to summon an `implicit` `Shrink`?

I ask, because I don't really think the implicit part is a good idea (ie `withShrink(shrink: Shrink[A])` would be what I would suggest). But I don't have enough data points to know what people are actually doing with Shrink. I only have my biases. :)

Reply:
> I ask, because I don't really think the implicit part is a good idea (ie withShrink(shrink: Shrink[A]) would be what I would suggest). But I don't have enough data points to know what people are actually doing with Shrink. I only have my biases. :)

I tend to agree here, I don't think I would be using the implicitness here very often, but sometimes it might come in handy (say I have my own type with `implicit Shrink[MyType]`) and then I do `Gen.listOf(arbitrary[MyType]).withShrink` then shrinking would be automatically propagated to the list generator, but I could still provide explicit shrinkers for other generators.

Reply:
Explicit shrinks (ie. forAllShrink with arity > 1) would be extremely useful for me.

My workflow for working around this issue has always been to write Gen and Shrink instances in parallel, and to wrap them (case class AsciiString(string: String), Gen[AsciiString], Shrink[AsciiString]) if there is ambiguity (because that's less annoying than passing in the Pretty instances).

@charleso 's case class GenShrink[A](gen: Gen[A], shrink: Shrink[A]) looks perfect for this, especially .filter could pretty easily operate on both sides.

Reply:
@woparry Thanks for the comment, it's good to hear from someone who is using this.

Just curious, any reason you don't provide an implicit `Arbitrary` to make that `Gen` as the global/implicit version? Is it just that `Gen` works with the global `Shrink`, or something else?